### PR TITLE
Structural gas estimator.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by arkpar** [comment]
clang build fails

File: libsolidity/ASTPrinter.cpp [hunk 1]
```diff
@@ -30,8 +30,10 @@ namespace dev
 namespace solidity
 {
 
-ASTPrinter::ASTPrinter(ASTNode const& _ast, string const& _source):
-	m_indentation(0), m_source(_source), m_ast(&_ast)
+ASTPrinter::ASTPrinter(ASTNode const& _ast,
```
Comment (LianaHus) [inline libsolidity/ASTPrinter.cpp]: _asm to the new line

**Comment by chriseth** [comment]
Build is fixed.

File: libsolidity/StructuralGasEstimator.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		gasCosts[&_node][0] = gasCosts[&_node][1] = particularCosts[_node.getLocation()];
+		return true;
+	};
+	auto onEdge = [&](ASTNode const& _parent, ASTNode const& _child)
+	{
+		gasCosts[&_parent][1] += gasCosts[&_child][1];
+	};
+	ASTReduce folder(onNode, onEdge);
+	for (ASTNode const* ast: _ast)
+		ast->accept(folder);
```
Comment (LianaHus) [inline libsolidity/StructuralGasEstimator.cpp]: please add a check  weather the pointer is not null

Comment (chriseth) [inline libsolidity/StructuralGasEstimator.cpp]: Added assertions.

**Pull request merged**